src/librustc_lint/builtin.rs

```diff
@@ -953,21 +952,31 @@ impl LintPass for NonSnakeCase {
         lint_array!(NON_SNAKE_CASE)
     }
 
+    fn check_crate(&mut self, cx: &Context, cr: &ast::Crate) {
+        let attr_crate_name = cr.attrs.iter().find(|at| at.check_name("crate_name"))
+                                      .and_then(|at| at.value_str());
+        if let Some(ref name) = cx.tcx.sess.opts.crate_name {
+            self.check_snake_case(cx, "crate", name, None);
+        } else if let Some(ref name) = attr_crate_name {
+            self.check_snake_case(cx, "crate", name, None);
```

Comment:
In this branch we have an attribute, so perhaps the span of the attribute could e passed?